strings/boyer_moore_good_suffix.cpp

```diff
@@ -0,0 +1,119 @@
+/**
+ * \file
+ * \brief The [Boyer-Moore Algorithm](Good Suffix Rule)(https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm) for
+ * finding a pattern within a piece of text with complexity O(m) in processing + o(mn)(worst case)
+ */
+ #include <iostream> 
+ #include<vector>
+// preprocessing for strong good suffix rule
+// A border is a substring which is both proper suffix and proper prefix. 
+// bpos is border position array,  Each entry bpos[i] contains the starting index of border for suffix starting at index i in given pattern P.
+// shift is an integer array,  Each entry shift[i] contain the distance pattern will shift if mismatch occur at position i-1. 
+
+//using struct to return multiple object
+struct contain{
+    std::vector<int> shift;
+    std::vector<int> bpos;
+};
+contain preprocess_strong_suffix(std::vector<int> shift, std::vector<int> bpos, std::string pat, int m) 
```

Comment:
Missing function documentation (see #962 as a reference for how to document functions). Check other parts of the code.